File: bindings/zig/src/lean_multisig_test.zig
const std = @import("std");
const lms = @import("lean_multisig.zig");

test "signs, aggregates, restores, and verifies" {
    const allocator = std.testing.allocator;
    try lms.setup();
    const claim = lms.Claim{ .message = [_]u8{42} ** 32, .slot = 100 };

    var alice = try lms.SecretKey.fromSeed([_]u8{1} ** 32, 100, 115);
    defer alice.deinit();
    var bob = try lms.SecretKey.fromSeed([_]u8{2} ** 32, 100, 115);
    defer bob.deinit();

    var alice_signature = try alice.sign(claim);
    defer alice_signature.deinit();
    var bob_signature = try bob.sign(claim);
    defer bob_signature.deinit();
    const signatures = [_]*const lms.Signature{ &alice_signature, &bob_signature };
    var combined = try lms.aggregate(allocator, &signatures, claim);
    defer combined.deinit();

    const signers = [_]lms.PublicKey{
        try alice.publicKey(allocator),
        try bob.publicKey(allocator),
    };
    try std.testing.expect(try combined.verify(&signers, claim));

    const encoded = try combined.toBytes(allocator);
    defer allocator.free(encoded);
    var restored = try lms.Signature.fromBytes(encoded, claim, &signers);
    defer restored.deinit();
    try std.testing.expect(try restored.verify(&signers, claim));
}

test "merges claims and restores the proof" {
    const allocator = std.testing.allocator;
    try lms.setup();
    const first = lms.Claim{ .message = [_]u8{3} ** 32, .slot = 100 };
    const second = lms.Claim{ .message = [_]u8{4} ** 32, .slot = 101 };
    var key = try lms.SecretKey.fromSeed([_]u8{5} ** 32, 100, 115);
    defer key.deinit();
    var first_signature = try key.sign(first);
    defer first_signature.deinit();
    var second_signature = try key.sign(second);
    defer second_signature.deinit();
    const signatures = [_]*const lms.Signature{ &first_signature, &second_signature };
    var proof = try lms.mergeClaims(allocator, &signatures);
    defer proof.deinit();

    const public_key = try key.publicKey(allocator);
    const signer_set = [_]lms.PublicKey{public_key};
    const groups = [_]lms.ClaimSigners{
        .{ .claim = first, .signers = &signer_set },
        .{ .claim = second, .signers = &signer_set },
    };
    try std.testing.expect(try proof.verifyClaims(allocator, &groups));

    const encoded = try proof.toBytes(allocator);
    defer allocator.free(encoded);
    var restored = try lms.MultiClaimProof.fromBytes(allocator, encoded, &groups);
    defer restored.deinit();
    var actual = try restored.verifiedClaims(allocator);
    defer actual.deinit(allocator);
    try std.testing.expectEqual(@as(usize, 2), actual.items.len);
    try std.testing.expect(try restored.verifyClaims(allocator, actual.items));
}

File: bindings/go/leanmultisig_test.go
package leanmultisig

import (
	"reflect"
	"testing"
)

func TestSignAggregateAndRestore(t *testing.T) {
	if err := Setup(); err != nil {
		t.Fatal(err)
	}
	claim := Claim{Message: bytes32(42), Slot: 100}

	alice, err := SecretKeyFromSeed(bytes32(1), 100, 115)
	if err != nil {
		t.Fatal(err)
	}
	defer alice.Close()
	bob, err := SecretKeyFromSeed(bytes32(2), 100, 115)
	if err != nil {
		t.Fatal(err)
	}
	defer bob.Close()

	aliceSignature, err := alice.Sign(claim)
	if err != nil {
		t.Fatal(err)
	}
	defer aliceSignature.Close()
	bobSignature, err := bob.Sign(claim)
	if err != nil {
		t.Fatal(err)
	}
	defer bobSignature.Close()
	aggregate, err := Aggregate([]*Signature{aliceSignature, bobSignature}, claim)
	if err != nil {
		t.Fatal(err)
	}
	defer aggregate.Close()

	encoded, err := aggregate.Bytes()
	if err != nil {
		t.Fatal(err)
	}
	alicePublicKey := mustPublicKey(t, alice)
	bobPublicKey := mustPublicKey(t, bob)
	restored, err := SignatureFromBytes(encoded, claim, [][32]byte{alicePublicKey, bobPublicKey})
	if err != nil {
		t.Fatal(err)
	}
	defer restored.Close()

	valid, err := Verify(restored, [][32]byte{alicePublicKey, bobPublicKey}, claim)
	if err != nil {
		t.Fatal(err)
	}
	if !valid {
		t.Fatal("restored aggregate must verify")
	}
}

func TestMergeClaims(t *testing.T) {
	if err := Setup(); err != nil {
		t.Fatal(err)
	}
	first := Claim{Message: bytes32(3), Slot: 100}
	second := Claim{Message: bytes32(4), Slot: 101}
	key, err := SecretKeyFromSeed(bytes32(5), 100, 115)
	if err != nil {
		t.Fatal(err)
	}
	defer key.Close()
	firstSignature, err := key.Sign(first)
	if err != nil {
		t.Fatal(err)
	}
	defer firstSignature.Close()
	secondSignature, err := key.Sign(second)
	if err != nil {
		t.Fatal(err)
	}
	defer secondSignature.Close()
	proof, err := MergeClaims([]*Signature{firstSignature, secondSignature})
	if err != nil {
		t.Fatal(err)
	}
	defer proof.Close()

	publicKey := mustPublicKey(t, key)
	expected := []ClaimSigners{
		{Claim: first, Signers: [][32]byte{publicKey}},
		{Claim: second, Signers: [][32]byte{publicKey}},
	}
	valid, err := VerifyClaims(proof, expected)
	if err != nil {
		t.Fatal(err)
	}
	if !valid {
		t.Fatal("multi-claim proof must verify")
	}
	encoded, err := proof.Bytes()
	if err != nil {
		t.Fatal(err)
	}
	restored, err := MultiClaimProofFromBytes(encoded, expected)
	if err != nil {
		t.Fatal(err)
	}
	defer restored.Close()
	actual, err := VerifiedClaims(restored)
	if err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(expected, actual) {
		t.Fatalf("unexpected verified claims: %#v", actual)
	}
}

func mustPublicKey(t *testing.T, key *SecretKey) [32]byte {
	t.Helper()
	publicKey, err := key.PublicKey()
	if err != nil {
		t.Fatal(err)
	}
	return publicKey
}

func bytes32(value byte) [32]byte {
	var bytes [32]byte
	for index := range bytes {
		bytes[index] = value
	}
	return bytes
}
